cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/ShardedHBaseQueueStrategy.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright © 2014-2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.transaction.queue.hbase;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.data2.queue.ConsumerConfig;
+import co.cask.cdap.data2.queue.ConsumerGroupConfig;
+import co.cask.cdap.data2.queue.DequeueStrategy;
+import co.cask.cdap.data2.queue.QueueEntry;
+import co.cask.cdap.data2.transaction.queue.QueueEntryRow;
+import co.cask.cdap.data2.transaction.queue.QueueScanner;
+import co.cask.cdap.hbase.wd.DistributedScanner;
+import com.google.common.base.Function;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.util.Threads;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An implementation of {@link HBaseQueueStrategy} with sharded keys.
+ *
+ * The row key has structure of:
+ *
+ * <pre>
+ * {@code
+ *
+ * row_key = <shard> <queue_prefix> <write_pointer> <counter>
+ * shard = <salt> <consumer_group_id> <consumer_instance_id>
+ * salt = First byte of MD5 of <consumer_group_id>, <consumer_instance_id> and <queue_name>
+ * consumer_group_id = 8 bytes long value of the target consumer group or 0 if it is FIFO
```

Comment:
I guess because in fifo it doesn't matter how many consumer instances there are, so we just share same row set between all consumer groups that use fifo?